geozero/src/proj.rs

```diff
@@ -0,0 +1,139 @@
+    PropertyProcessor,
+};
+
+use delegate::delegate;
+use proj::Proj;
+
+use std::rc::Rc;
+
+/// ```
+/// use proj::Proj;
+/// use geozero::{ToProjected, ProcessToJson, wkt::WktStr};
+///
+/// // Convert from [NAD 83 US Survey Feet](https://epsg.io/2230) to [NAD 83 Meters](https://epsg.io/26946) Using EPSG Codes
+/// let from = "EPSG:2230";
+/// let to = "EPSG:26946";
+/// let proj = Proj::new_known_crs(from, to, None).unwrap();
+///
+/// let wkt = WktStr("POINT (4760096.421921 3744293.729449)");
+/// let expected = r#"{"type": "Point", "coordinates": [1450880.2910605022,1141263.0111604782]}"#;
+/// assert_eq!(expected, &wkt.to_projected(proj).to_json().unwrap());
+/// ```
+pub trait ToProjected: GeozeroDatasource + Sized {
+    fn to_projected(self, proj: Proj) -> ProjWrappedDataSource<Self>;
+}
+
+impl<Input: GeozeroDatasource> ToProjected for Input {
+    fn to_projected(self, proj: Proj) -> ProjWrappedDataSource<Input> {
+        ProjWrappedDataSource {
+            proj: Rc::new(proj),
+            input: self,
+        }
+    }
+}
+
+pub struct ProjWrappedDataSource<Input: GeozeroDatasource> {
+    proj: Rc<Proj>,
+    input: Input,
+}
+
+struct ProjPreprocessor<'a> {
+    proj: Rc<Proj>,
+    output: &'a mut dyn FeatureProcessor,
+}
+
+impl<Input: GeozeroDatasource> GeozeroDatasource for ProjWrappedDataSource<Input> {
+    fn process<P: FeatureProcessor>(&mut self, processor: &mut P) -> Result<()> {
+        let mut wrapped = ProjPreprocessor {
+            proj: self.proj.clone(),
+            output: processor,
+        };
+        self.input.process(&mut wrapped)
+    }
+}
+
+impl FeatureProcessor for ProjPreprocessor<'_> {
+    delegate! {
+         to self.output {
+             fn dataset_begin(&mut self, name: Option<&str>) -> Result<()>;
+             fn dataset_end(&mut self) -> Result<()>;
+             fn feature_begin(&mut self, idx: u64) -> Result<()>;
+             fn feature_end(&mut self, idx: u64) -> Result<()>;
+             fn properties_begin(&mut self) -> Result<()>;
+             fn properties_end(&mut self) -> Result<()>;
+             fn geometry_begin(&mut self) -> Result<()>;
+             fn geometry_end(&mut self) -> Result<()>;
+        }
+    }
+}
+
+impl PropertyProcessor for ProjPreprocessor<'_> {
+    delegate! {
+        to self.output {
+            fn property(&mut self, idx: usize, name: &str, value: &ColumnValue) -> Result<bool>;
+        }
+    }
+}
+
+impl GeomProcessor for ProjPreprocessor<'_> {
+    fn coordinate(
+        &mut self,
+        x: f64,
+        y: f64,
+        z: Option<f64>,
+        m: Option<f64>,
+        t: Option<f64>,
+        tm: Option<u64>,
+        idx: usize,
+    ) -> Result<()> {
+        let (x, y) = self.proj.convert((x, y)).unwrap();
+        self.output.coordinate(x, y, z, m, t, tm, idx)
+    }
+
+    fn xy(&mut self, x: f64, y: f64, idx: usize) -> Result<()> {
+        let (x, y) = self.proj.convert((x, y)).unwrap();
+        self.output.xy(x, y, idx)
+    }
```

Comment:
This is the crux - we transform the points before passing them in to the "next" processor in the pipeline.

We only "need" these two methods `coordinate` / `xy`, but we have to explicitly delegate our other impls to the "output" processor, otherwise we'd inherit the "no-op" implementations for the GeomProcessor trait.

Effectively, I'm going for some kind of interceptor pattern here.

But it's very verbose. And also, it'd be easy to break if we added a new method to GeomProcessor/FeatureProcessor/PropertyProcessor and failed to update these long list of delegations below.

Reply:
Sorry, I'm currently short on time and leaving for Easter holidays soon. My only input would be that maybe a delegation pattern like https://github.com/flatgeobuf/flatgeobuf/blob/845111d2ba6765c8f43379877eb1f3440b78d4b6/src/rust/src/file_writer.rs#L302 could work here as well.

Reply:
> maybe a delegation pattern

If I understand correctly, that's exactly what I'm doing - I did use the `delegate` crate macro to make it a *little* more concise, but it seems like a *lot* of boilerplate.

I'm searching for, but have so far failed to find, a better way to do this.